src/HeapedVectors.jl

```diff
@@ -0,0 +1,156 @@
+__precompile__()
+
+struct HeapedVector{T, F<:Function}  
+    data::Vector{T}
+    by::F
+    function HeapedVector(v::Vector{T}, by::F) where {T, F}
+        new{T, F}(heaping(v, by), by)
+    end
+end
+      
+HeapedVector(data::Vector{T}) where {T} = HeapedVector(data, identity)  
+
+       
+function heaping(v, by) 
+    ar = []
+    for i = 1:length(v) 
+        insert!(ar, i, v[i])
+        floatup(length(ar), ar, by)
+    end
+    return ar
+end
+
+function floatup(index, ar, by)	
+    par = convert(Int, floor(index/2))
+    if index <= 1
+        return ar 
+    end
+    if by(ar[index]) < by(ar[par])
+        temp = ar[par]
+        ar[par] = ar[index]
+        ar[index] = temp
+    end
+    floatup(par, ar, by)
+end
+
+
+function push!(v::HeapedVector{T}, x::T) where {T}
+    insert!(v.data, length(v.data)+1, x)
+    floatup(length(v.data), v.data, v.by)
+    return v
+end
+
+
+isempty(v::HeapedVector) = isempty(v.data)
+
+
+function popfirst!(v::HeapedVector{T}) where {T}
+    if length(v.data) == 0
+         return 
+    end
+	
+    if length(v.data) > 2
+        temp = v.data[length(v.data)]
+    	v.data[length(v.data)] = v.data[1]
+    	v.data[1] = temp
+    	Min = pop!(v.data)
+    	bubbledown(1, v::HeapedVector{T})
+    
+    elseif length(v.data) == 2
+    	temp = v.data[length(v.data)]
+    	v.data[length(v.data)] = v.data[1]
+    	v.data[1] = temp
+    	Min = pop!(v.data)
+    else
+    	Min = pop!(v.data)	
+    end
+    return Min
+end
+
+
+function bubbledown(index, v::HeapedVector{T}) where{T}
+    left = index*2
+    right = index*2+1
+    smallest = index
+
+    if length(v.data)+1 > left && v.by(v.data[smallest]) > v.by(v.data[left])
+        smallest = left
+    end
+
+    if length(v.data)+1 > right && v.by(v.data[smallest]) > v.by(v.data[right])
+	   smallest = right
+    end
+
+    if smallest != index
+	   temp = v.data[index]
+	   v.data[index] = v.data[smallest]
+	   v.data[smallest] = temp
+	   bubbledown(smallest, v)
+    end
+end
+
+function cutoff(A::HeapedVector{T}, index::Int, x::T, ran) where{T} 
+    if index > length(A.data)
+        return A
+    end
+
+    if A.by(A.data[index]) >= A.by(x)
+        A.data[index] = (x[1], ran[1])
+    end
+    
+    cutoff(A, index*2, x, ran)
+    cutoff(A, index*2+1, x, ran)
+
+end
+
+function delete_elements!(A::HeapedVector{T}, x::T) where{T} 
+    if length(A.data)==0
+        return A
+    end
+    ran=rand(typeof(1.0), 1)
```

Comment:
Why is there a random number in here??

Reply:
Actually, according to logic of the function `delete_elements!` I am  first replacing the elements which should be remove by some random element and in the end deleting all those random element.

Reply:
This is a bad idea. E.g. it will delete a value that I don't want to be deleted if it happens to be equal to this random element. Also, the random number function will not work for all types `T`.

Why can't you directly delete the values that need deleting?